synapse/replication/tcp/handler.py

```diff
@@ -401,12 +403,27 @@ def on_USER_IP(
     ) -> Optional[Awaitable[None]]:
         user_ip_cache_counter.inc()
 
-        if self._is_master:
+        if self._is_master or self._is_background_worker:
             return self._handle_user_ip(cmd)
         else:
             return None
 
     async def _handle_user_ip(self, cmd: UserIpCommand) -> None:
+        """
+        Handles a User IP, branching depending on whether we are the main process
+        and/or the background worker.
+        """
+        if self._is_master:
+            await self._handle_user_ip_as_master(cmd)
+
+        if self._is_background_worker:
+            await self._handle_user_ip_as_background_worker(cmd)
```

Comment:
I think it would be clearer to inline `_handle_user_ip` into `on_USER_IP`? Otherwise its a bit confusing

Reply:
Only concern with that is that it will mean that uninvolved workers end up creating and scheduling coroutines as background processes as well, whereas they didn't before.
I don't know how much that worries us, but since the code was already structured this way for some reason, I didn't want to remove it without understanding.

Reply:
I'm confused, I'm not talking about a functional change here? `_handle_user_ip` is only called by `on_USER_IP`?

What's confusing me really is that the `self._is_master or self._should_insert_client_ips` a) looks odd by itself, and b) is entirely redundant given `_handle_user_ip` then does the exact same checks. You may as well just always call `self._handle_user_ip` in `on_USER_IP`, at which point having it as a separate function seems a bit silly?

Or am I missing something here?

Reply:
`on_USER_IP` returns either `None` or an awaitable. If it's an awaitable, it goes through extra stuff to schedule it as a background process; if it's just `None` then nothing happens.

Notably, the function can't (easily/nicely) be inlined and still preserve this behaviour of returning `None` if there's nothing to do, as I'd have to make `on_USER_IP` an `async def` in order to run potentially two coroutines back-to-back.

It may not be a big deal, but I wasn't sure and since replication overhead is pretty sensitive, I was a bit uncomfortable taking a leap of faith.

Reply:
OH. Right, with you now. How obscure.

Reply:
Yes, and it took me forever where to find the code that calls `on_%s` because the name is dynamically constructed and I forgot that we like `%s`-formatting and was searching for `"on_"` :')

Reply:
Can you add a comment in `on_USER_IP` about the whole awaitable thing please?

Reply:
Done — I also inlined the two specific `handle_user_ip` functions, since looking at them now, they're so short it wasn't buying any clarity.